compiler/resolution/functionResolution.cpp

```diff
@@ -4926,15 +4926,21 @@ static void resolveInitField(CallExpr* call) {
   if (!fs)
     INT_FATAL(call, "bad initializer set field primitive");
 
-  Type* srcType = srcExpr->symbol()->type;
+  Symbol* srcSym = srcExpr->symbol();
+  Type* srcType = srcSym->type;
   if (srcType == dtUnknown)
     INT_FATAL(call, "Unable to resolve field type");
 
   if (fs->hasFlag(FLAG_PARAM)) {
     if (isLegalParamType(srcType) == false) {
       USR_FATAL_CONT(fs, "'%s' is not of a supported param type", fs->name);
     }
-    if (srcExpr->symbol()->isParameter() == false) {
+
+    // Check against the symbol's Immediate is required because the symbol is
+    // likely a temporary, which generally will suppress an error if
+    // initialized from a non-param.
+    else if (srcSym->isParameter() == false ||
+            (isVarSymbol(srcSym) && getImmediate(srcSym) == NULL)) {
```

Comment:
This makes me nervous.  It basically means `hasFlag(FLAG_PARAM)` is not a valid check at this point in time, because srcSym->isParameter should dispatch to VarSymbol's isParameter, which checks for either the flag or the presence of an immediate.

I would prefer tracking down where the temporary fails to get checked.  I suspect it is being created with `FLAG_MAYBE_PARAM` instead of `FLAG_PARAM`, when we should know at the time it is being used to create a param variable.

Reply:
> I suspect it is being created with FLAG_MAYBE_PARAM instead of FLAG_PARAM, when we should know at the time it is being used to create a param variable.

Sorry, ignore that.  I still think we should track down where the temporary suppresses the error

Reply:
I'm not as sure because a lot of parts of the compiler ignore or postpone checking of temporaries because of our early lowering. I think it would be easy to detect when initializing the temporary with a non-param, but generating a useful error message would be more difficult.

Reply:
Ah, because it would have lost the name of the original variable at that point?

Reply:
Yeah, and then we'd have to examine the surrounding AST to figure out the context.

Reply:
Ugh.  Maybe just add a comment that this is kinda gross?

Reply:
Thanks!  (also, I think you know this, but in case it wasn't clear, I'm annoyed at the implementation causing these choices, not you for having to make it)